internal/controller/servicetemplate_controller.go

```diff
@@ -0,0 +1,580 @@
+	var err error
+
+	status := kcm.ServiceTemplateStatus{
+		TemplateStatusCommon: kcm.TemplateStatusCommon{
+			TemplateValidationStatus: kcm.TemplateValidationStatus{},
+			ObservedGeneration:       template.Generation,
+		},
+	}
+
+	switch ref.Kind {
+	case "Secret":
+		secret := &corev1.Secret{}
+		err = r.Get(ctx, key, secret)
+		if err != nil {
+			err = fmt.Errorf("failed to get referred Secret %s: %w", key, err)
+			break
+		}
+		status.SourceStatus, err = r.sourceStatusFromLocalObject(secret)
+		if err != nil {
+			return fmt.Errorf("failed to get source status from Secret %s: %w", key, err)
+		}
+	case "ConfigMap":
+		configMap := &corev1.ConfigMap{}
+		err = r.Get(ctx, key, configMap)
+		if err != nil {
+			err = fmt.Errorf("failed to get referred ConfigMap %s: %w", key, err)
+			break
+		}
+		status.SourceStatus, err = r.sourceStatusFromLocalObject(configMap)
+		if err != nil {
+			return fmt.Errorf("failed to get source status from ConfigMap %s: %w", key, err)
+		}
+	case "GitRepository":
+		gitRepository := &sourcev1.GitRepository{}
+		err = r.Get(ctx, key, gitRepository)
+		if err != nil {
+			err = fmt.Errorf("failed to get referred GitRepository %s: %w", key, err)
+			break
+		}
+		status.SourceStatus, err = r.sourceStatusFromFluxObject(gitRepository)
+		if err != nil {
+			return fmt.Errorf("failed to get source status from GitRepository %s: %w", key, err)
+		}
+		conditions := make([]metav1.Condition, len(gitRepository.Status.Conditions))
+		copy(conditions, gitRepository.Status.Conditions)
+		status.SourceStatus.Conditions = conditions
+	case "Bucket":
+		bucket := &sourcev1.Bucket{}
+		err = r.Get(ctx, key, bucket)
+		if err != nil {
+			err = fmt.Errorf("failed to get referred Bucket %s: %w", key, err)
+			break
+		}
+		status.SourceStatus, err = r.sourceStatusFromFluxObject(bucket)
+		if err != nil {
+			return fmt.Errorf("failed to get source status from Bucket %s: %w", key, err)
+		}
+		conditions := make([]metav1.Condition, len(bucket.Status.Conditions))
+		copy(conditions, bucket.Status.Conditions)
+		status.SourceStatus.Conditions = conditions
+	case "OCIRepository":
+		ociRepository := &sourcev1beta2.OCIRepository{}
+		err = r.Get(ctx, key, ociRepository)
+		if err != nil {
+			err = fmt.Errorf("failed to get referred OCIRepository %s: %w", key, err)
+			break
+		}
+		status.SourceStatus, err = r.sourceStatusFromFluxObject(ociRepository)
+		if err != nil {
+			return fmt.Errorf("failed to get source status from OCIRepository %s: %w", key, err)
+		}
+		conditions := make([]metav1.Condition, len(ociRepository.Status.Conditions))
+		copy(conditions, ociRepository.Status.Conditions)
+		status.SourceStatus.Conditions = conditions
+	default:
+		err = fmt.Errorf("unsupported source kind %s", ref.Kind)
+	}
+
+	if err != nil {
+		status.TemplateValidationStatus.Valid = false
+		status.TemplateValidationStatus.ValidationError = err.Error()
+	} else {
+		status.TemplateValidationStatus.Valid = true
+	}
+	template.Status = status
```

Comment:
I would say this is the right place for a `defer`, something like

```go
func (r *receiver) methodName(arguments ...struct{}) (err error) {
  defer func() {
    if err != nil { failed; } else { valid; }
    template.Status = status
  }
  
  switch ref.Kind {
  case "Kind1": 
    if err := r.Client.Get(); err != nil { return fmt.Errorf("wrapme: %w", err);} ;
  default:
    return fmt.Errorf("unexpected")
}
```

that way it would be easier to read and no need to preserve the global (to the method) scope of the `err`, no need in the `break` statements, etc. Just plain `return` from each case

Let me know what you think